src/main/java/htsjdk/variant/vcf/VCFHeader.java

```diff
@@ -98,135 +92,261 @@
+                    !requestedVCFHeaderVersion.equals(embeddedHeaderVersion)) {
+                final String message = String.format("VCFHeader metadata version (%s) is inconsistent with requested version (%s). " +
+                                "Falling back to metadata version.",
+                                embeddedHeaderVersion,
+                                requestedVCFHeaderVersion);
+                if (VCFUtils.getStrictVCFVersionValidation()) {
+                    throw new IllegalArgumentException(message);
+                }
+                if (VCFUtils.getVerboseVCFLogging()) {
+                    logger.warn(message);
+                }
+            }
+            return embeddedHeaderVersion;
+        }
+    }
+
+   /**
+    * Get the header version for this header.
+    * @return the VCFHeaderVersion for this header.
+    */
+    public VCFHeaderVersion getHeaderVersion() {
+        return vcfVersion;
+    }
+
+    /**
+     * Set the version header for this class.
+     *
+     * Validates all metadata line to ensure they conform to the target header version (i.e, if the metadata lines
+     * contain a ##fileformat line that specifies a version that is different than the {@code newVCFVersion}.
+     *
+     * @param newVCFVersion
+     */
+    public void setHeaderVersion(final VCFHeaderVersion newVCFVersion) {
+        Utils.nonNull(newVCFVersion, "A non-null VCFHeaderVersion must be provided");
+        if (!newVCFVersion.equals(vcfVersion)) {
+            logger.warn(String.format("Changing VCFHeader version from %s to %s",
+                    vcfVersion.getVersionString(),
+                    newVCFVersion.getVersionString()));
+            // TODO: This can cause failures in code that used to succeed (i.e. Picard LiftOverVcf tests fail
+            // if they're not modified to remove the embedded version line) since we now retain ##fileformat
+            // lines in the metadata list; the validation code recognizes and validates the embedded ##fileformat
+            // lines against the new version, and throws if they conflict.
+            //
+            // We might want to add a removeHeaderLine method to VCFHeader so that consumers with this problem
+            // such as LiftOverVcf can first manually remove the embedded fileformat line (currently you'd have
+            // to create a new header to achieve that).
+            mMetaData.validateMetaDataLines(newVCFVersion);
+        }
+    }
+
+    /**
+     * Adds a new line to the VCFHeader. If there is an existing header line of the
+     * same type with the same key, and the header version is pre-4.3, the new line is added
+     * using a modified key to make it unique for BWC; otherwise (in strict validation mode), an
+     * exception is thrown since duplicates are not allowed.
+     *
+     * @param headerLine header line to attempt to add
+     */
+    public void addMetaDataLine(final VCFHeaderLine headerLine) {
```

Comment:
Maybe a method to override a meta data line will be useful, to being able to remove the previous associated key. It can return the previous line to check if it was present...